src/common/acl_util.c

```diff
@@ -0,0 +1,331 @@
+/*
+ * (C) Copyright 2019 Intel Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+ * The Government's rights to use, modify, reproduce, release, perform, display,
+ * or disclose this software are subject to the terms of the Apache License as
+ * provided in Contract No. 8F-30005.
+ * Any reproduction of computer software, computer software documentation, or
+ * portions thereof marked with this legend must also reproduce the markings.
+ */
+
+/**
+ * Utility functions that may be used when interacting with Access Control
+ * Lists
+ */
+
+#include <daos_security.h>
+#include <gurt/common.h>
+#include <gurt/debug.h>
+#include <pwd.h>
+#include <grp.h>
+
+static size_t DEFAULT_BUF_LEN = 1024;
+
+enum validity_state {
+	STATE_START,
+	STATE_NAME,
+	STATE_DOMAIN,
+	STATE_DONE,
+	STATE_INVALID
+};
+
+static enum validity_state
+next_state(enum validity_state current_state, char ch)
+{
+	switch (current_state) {
+	case STATE_START:
+		if (ch == '@')
+			return STATE_INVALID;
+		return STATE_NAME;
+	case STATE_NAME:
+		if (ch == '@')
+			return STATE_DOMAIN;
+		if (ch == '\0')
+			return STATE_INVALID;
+		break;
+	case STATE_DOMAIN:
+		if (ch == '\0')
+			return STATE_DONE;
+		if (ch == '@')
+			return STATE_INVALID;
+		break;
+	case STATE_DONE:
+	case STATE_INVALID:
+		break;
+	default:
+		D_ASSERTF(false, "Invalid state: %d\n", current_state);
+	}
+
+	return current_state;
+}
+
+bool
+daos_acl_principal_is_valid(const char *name)
+{
+	size_t			len;
+	size_t			i;
+	enum validity_state	state = STATE_START;
+
+	if (name == NULL) {
+		D_ERROR("Name was NULL\n");
+		return false;
+	}
+
+	len = strnlen(name, DAOS_ACL_MAX_PRINCIPAL_BUF_LEN);
+	if (len == 0 || len > DAOS_ACL_MAX_PRINCIPAL_LEN) {
```

Comment:
do you need to check if it is greater than the Max? surely using strnlen means that would never be the case

Reply:
I added the DAOS_ACL_MAX_PRINCIPAL_BUF_LEN constant to include an extra character for the null terminator. For strnlen I'm intentionally using the principal's max buf len (rather than max string len) to allow it to detect a string that's too long.